### PR TITLE
CodeArtifact encryption_key argument should be optional+computed

### PR DESCRIPTION
**Pull request by j3parker**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_codeartifact_domain: Change encryption_key to an optional+computed attribute
```
Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCodeArtifactDomain*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCodeArtifactDomain_* -timeout 120m
=== RUN   TestAccAWSCodeArtifactDomainPermissionsPolicy_basic
=== PAUSE TestAccAWSCodeArtifactDomainPermissionsPolicy_basic
=== RUN   TestAccAWSCodeArtifactDomainPermissionsPolicy_owner
=== PAUSE TestAccAWSCodeArtifactDomainPermissionsPolicy_owner
=== RUN   TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears
=== PAUSE TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears
=== RUN   TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears_domain
=== PAUSE TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears_domain
=== RUN   TestAccAWSCodeArtifactDomain_basic
=== PAUSE TestAccAWSCodeArtifactDomain_basic
=== RUN   TestAccAWSCodeArtifactDomain_defaultencryptionkey
=== PAUSE TestAccAWSCodeArtifactDomain_defaultencryptionkey
=== RUN   TestAccAWSCodeArtifactDomain_disappears
=== PAUSE TestAccAWSCodeArtifactDomain_disappears
=== CONT  TestAccAWSCodeArtifactDomainPermissionsPolicy_basic
=== CONT  TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears_domain
=== CONT  TestAccAWSCodeArtifactDomain_basic
=== CONT  TestAccAWSCodeArtifactDomain_disappears
=== CONT  TestAccAWSCodeArtifactDomain_defaultencryptionkey
=== CONT  TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears
=== CONT  TestAccAWSCodeArtifactDomainPermissionsPolicy_owner
2020/10/14 13:48:17 [DEBUG] Waiting for state to become: [success]
=== CONT  TestAccAWSCodeArtifactDomain_defaultencryptionkey
    resource_aws_codeartifact_domain_test.go:104: Step 1/2 error: Error running apply:
        Error: error creating CodeArtifact Domain: InvalidParameter: 1 validation error(s) found.
        - minimum field size of 1, CreateDomainInput.EncryptionKey.



--- FAIL: TestAccAWSCodeArtifactDomain_defaultencryptionkey (32.34s)
--- PASS: TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears (40.24s)
--- PASS: TestAccAWSCodeArtifactDomainPermissionsPolicy_disappears_domain (44.32s)
--- PASS: TestAccAWSCodeArtifactDomain_basic (65.13s)
--- PASS: TestAccAWSCodeArtifactDomainPermissionsPolicy_owner (75.98s)
--- PASS: TestAccAWSCodeArtifactDomain_disappears (81.95s)
--- PASS: TestAccAWSCodeArtifactDomainPermissionsPolicy_basic (100.26s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       100.306s
FAIL
GNUmakefile:26: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

### CODE REVIEW

File: aws/resource_aws_codeartifact_domain.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ func resourceAwsCodeArtifactDomain() *schema.Resource {
 			},
 			"encryption_key": {
 				Type:         schema.TypeString,
-				Required:     true,
+				Optional:     true,
```
Comment (j3parker) [inline aws/resource_aws_codeartifact_domain.go:36]: It is optional in the SDK: https://docs.amazonaws.cn/sdk-for-go/api/service/codeartifact/#CreateDomainInput
And in the API: https://docs.aws.amazon.com/codeartifact/latest/APIReference/API_CreateDomain.html#codeartifact-CreateDomain-request-encryptionKey

Comment (j3parker) [reply aws/resource_aws_codeartifact_domain.go:36]: CC @DrFaust92

Comment (DrFaust92) [reply aws/resource_aws_codeartifact_domain.go:36]: Hey @j3parker, 
I'll take a closer look but from what I recall the API validates that the kms Id field is not empty even if it's optional. That's why it's required in terraform.

Comment (j3parker) [reply aws/resource_aws_codeartifact_domain.go:36]: Here's what I get from the CLI:

```
$ aws configure set cli_history enabled

$ aws codeartifact create-domain --domain test
{
    "domain": {
        "name": "test",
        "owner": "(redacted)",
        "arn": "arn:aws:codeartifact:us-east-1:(redacted):domain/test",
        "status": "Active",
        "createdTime": "2020-10-09T10:33:21.559000-04:00",
        "encryptionKey": "arn:aws:kms:us-east-1:(redacted):key/(redacted)",
        "repositoryCount": 0,
        "assetSizeBytes": 0
    }
}

$ aws history show

AWS CLI command entered
at time: 2020-10-09 10:33:20.776
with AWS CLI version: aws-cli/2.0.49 Python/3.7.7 Windows/10 exe/AMD64
with arguments: ['codeartifact', 'create-domain', '--domain', 'test']

[0] API call made
at time: 2020-10-09 10:33:20.930
to service: codeartifact
using operation: CreateDomain
with parameters: {
    "domain": "test"
}

[0] HTTP request sent
at time: 2020-10-09 10:33:20.936
to URL: https://codeartifact.us-east-1.amazonaws.com/v1/domain?domain=test
with method: POST
with headers: {
    "Authorization": "(redacted)",
    "Content-Length": "0",
    "User-Agent": "aws-cli/2.0.49 Python/3.7.7 Windows/10 exe/AMD64 command/codeartifact.create-domain",
    "X-Amz-Date": "20201009T143320Z",
    "X-Amz-Security-Token": "(redacted)"
}
with body: There is no associated body

[0] HTTP response received
at time: 2020-10-09 10:33:21.466
with status code: 200
with headers: {
    "Connection": "keep-alive",
    "Content-Length": "354",
    "Content-Type": "application/json",
    "Date": "Fri, 09 Oct 2020 14:33:21 GMT",
    "x-amzn-RequestId": "(redacted)"
}
with body: {
    "domain": {
        "arn": "arn:aws:codeartifact:us-east-1:(redacted):domain/test",
        "assetSizeBytes": 0,
        "createdTime": 1602254001.559,
        "encryptionKey": "arn:aws:kms:us-east-1:(redacted):key/(redacted)",
        "name": "test",
        "owner": "(redacted)",
        "repositoryCount": 0,
        "s3BucketArn": "arn:aws:s3:::assets-(redacted)-us-east-1",
        "status": "Active"
    }
}
```

So it looks like it's truly optional (it's not needed even on the wire).
If it's not supplied it defaults to an AWS-managed CMK which is pretty typical for AWS services (they get created once per account, but only on-demand. AWS takes care of rotating them, they're free, you don't typically need to specify them, etc.)

**Comment by DrFaust92** [comment]
Maybe an issue for the go sdk specifically, thanks for checking this. I'll
try again  and open a ticket on go sdk repo if it fails again.

**Comment by DrFaust92** [comment]
@j3parker, can you also add computed attribute to kms key id? If none is passed and the API returns a key we need to handle diffs with computed.

Alsop please run acceptance tests.

Thanks again!

**Comment by j3parker** [comment]
(following the instructions to run the acceptance tests now)

**Comment by j3parker** [comment]
done -- attached the logs to the PR description, they passed.

**Comment by DrFaust92** [comment]
Hey @j3parker,

Can you add an acceptance test for this use case an update docs? Thanks.

**Comment by j3parker** [comment]
Sure thing.

**Comment by j3parker** [comment]
Done.

RE: documentation, I based the new sentence after the one for the [`kms_master_key_id` attribute for S3 buckets](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#kms_master_key_id).
Here's a screenshot in the UI to show that the key name alias:

image


RE: the test -- it's failing with this error:

```
=== CONT  TestAccAWSCodeArtifactDomain_defaultencryptionkey
    resource_aws_codeartifact_domain_test.go:104: Step 1/2 error: Error running apply:
        Error: error creating CodeArtifact Domain: InvalidParameter: 1 validation error(s) found.
        - minimum field size of 1, CreateDomainInput.EncryptionKey.
```

I assume that's what you're talking about with the golang SDK. Should I open an issue in that repo?
Btw -- I was pattern matching with this test case, let me know if it's what you wanted/if there is something more I should be doing.

File: website/docs/r/codeartifact_domain.html.markdown [hunk 2]
```diff
@@ -13,13 +13,8 @@ Provides a CodeArtifact Domain Resource.
 ## Example Usage
 
 ```hcl
-resource "aws_kms_key" "example" {
-  description = "domain key"
-}
-
 resource "aws_codeartifact_domain" "example" {
   domain         = "example"
-  encryption_key = aws_kms_key.example.arn
```
Comment (j3parker) [inline website/docs/r/codeartifact_domain.html.markdown:22]: Most resources seem to prefer the AWS managed keys by default in their docs. Let me know if you want a separate example for providing a custom key (the stuff in the red text here).

**Comment by j3parker** [comment]
Hmmm...

```go
package main                                                                                                                                                                                                     
import (
    "fmt"
    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/codeartifact"
)

func main() {
    sess, err := session.NewSession(&aws.Config{
        Region: aws.String("us-east-1")},
    )

    svc := codeartifact.New(sess)

    domain := "foo"
    res, err := svc.CreateDomain(&codeartifact.CreateDomainInput{
        Domain: &domain,
    })

    if err != nil {
        fmt.Println("Could not create domain", err)
        return
    }

    fmt.Println("stuff", res)
}
```

```
$ go run main.go
stuff {
  Domain: {
    Arn: "arn:aws:codeartifact:us-east-1:*****:domain/foo",
    AssetSizeBytes: 0,
    CreatedTime: 2020-10-14 18:46:53.39 +0000 UTC,
    EncryptionKey: "arn:aws:kms:us-east-1:*****:key/*****",
    Name: "foo",
    Owner: "*****",
    RepositoryCount: 0,
    Status: "Active"
  }
}
```

File: website/docs/r/codeartifact_domain.html.markdown (hunk 2 above)
Comment (DrFaust92) [reply website/docs/r/codeartifact_domain.html.markdown:22]: Looks good this way.

**Comment by j3parker** [comment]
Just fixing merge conflicts; hoping to look into this again today or Monday.

**Comment by ewbankkit** [comment]
Replaced by https://github.com/hashicorp/terraform-provider-aws/pull/17262.

**Pull request closed without merging**